### PR TITLE
New empty state for variations

### LINKED ISSUE #37147: [Enhancement]: New empty state for variations - no variations have been created yet
There's another empty state when the attributes with the `Used for variations` box are added, but the variations are not generated yet. There's an icon and copy that reads `No variations yet. Generate them from all added attributes or add a new variation manually.` The arrow points to the new "Generate variations" button.

Copy:
```
No variations yet. Generate them from all added attributes or add a new variation manually.
```

Buttons:
- `Generate variations`
"Create variations from all attributes" is no longer part of the dropdown. It is now a separate button labeled `Generate variations`. Clicking it generates the variations from all attributes with the `Used for variations` box checked.

- `Add manually`
"Add variations" is no longer part of the dropdown. It is now a separate button labeled `Add manually`. Clicking it adds a new variation form (the same as it does now).

Screenshot 2023-03-09 at 13 35 05

**Designs:**
`zs56kcy53K7GoGhgyjcQ2R-fi-211%3A34201`

```[tasklist]
### Acceptance criteria
- [ ] Users will see this screen when the product type is variable and there’s at least one attribute added with the `Used for variations` box checked.
- [ ] Add a `Generate variations` button that generates the variations from all attributes with the `Used for variations` box checked.
- [ ]  Add an `Add manually` button that adds a new variation form (the same as it does now).
- [ ] Hide the dropdown menu.
- [ ] Show the text: No variations yet. Generate them from all added attributes or add a new variation manually.
- [ ] Show the icons.
```

### PR DESCRIPTION
**Pull request by octaedro**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR adds another empty state when the attributes with the `Used for variations` box are added, but the variations are not generated yet. There's an image and copy that reads `No variations yet. Generate them from all added attributes or add a new variation manually.` There's also an arrow that points to the new "Generate variations" button.

Closes #37147.

### How to test the changes in this Pull Request:

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

1. Go to Products > Add New > Product data
2. Select `Variable product`.
3. Press `Attributes` and add (or create) a new Attribute.
4. Press `Variations` and verify that the new screen is visible.
5. Press `Generate variations` and verify that the variations are created as expected.
6. The old dropdown menu and the `Go` button are still visible.
7. Remove the created variations and verify that the new screen is visible again.
8. Press `Add manually` and verify that the form to create a variation is visible.

### Acceptance criteria
- [ ] Users will see this screen when the product type is variable and there’s at least one attribute added with the `Used for variations` box checked.
- [ ] Add a `Generate variations` button that generates the variations from all attributes with the `Used for variations` box checked.
- [ ]  Add an `Add manually` button that adds a new variation form (the same as it does now).
- [ ] Hide the dropdown menu.
- [ ] Show the text: No variations yet. Generate them from all added attributes or add a new variation manually.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [x] Have you included testing instructions?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js [hunk 1]
```diff
@@ -1459,6 +1500,12 @@ jQuery( function ( $ ) {
 					displaying_links.show();
 				}
 			}
+
+			// Show/hide variations empty state
+			setTimeout(
```
Comment (octaedro) [inline plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js]: The animation that removes the variation takes around 300ms. I added this timeout to prevent showing the empty state and the variation simultaneously (while it's being removed).

Comment (mattsherman) [reply plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js]: I think the root cause of the problem is that this function, `set_paginav()` is called before the ajax call in `load_variations()` finishes (I put some `console.log()` statements in to trace things).

Calling `show_hide_variation_empty_state()` from within here doesn't really feel correct, as this function "Sets the pagenav fields".

It would be good to consider if this could either be called from somewhere else, or even inside `load_variations()` after the ajax succeeds. Or, perhaps trigger an event from within `load_variations()` that we could hook up to.

**Review by mattsherman — CHANGES_REQUESTED** [review]
Hmm... I didn't realize more changes were being pushed to this PR while I was reviewing it. I left a comment about the use of the `setTimeout` in jQuery and did not review the SCSS yet because I figured that would change when you rebased/merged conflicts.

Ping me when it is ready for another review and I'll take another look!

**Comment by octaedro** [comment]
Hey @mattsherman,
I rebased this PR 2 hours ago. I didn't add anything else after that.

File: plugins/woocommerce/client/legacy/css/admin.scss [hunk 2]
```diff
@@ -1050,6 +1049,33 @@
 			}
 		}
 	}
+
+	.add-variation-container {
```
Comment (octaedro) [inline plugins/woocommerce/client/legacy/css/admin.scss:1069]: The linter fixed many things in this file, so it's hard to see the class I added.
The class is `.add-variation-container`.

File: plugins/woocommerce/assets/images/product_data/no-variation-background-image.svg
Comment (mattsherman) [inline plugins/woocommerce/assets/images/product_data/no-variation-background-image.svg:1]: Do we want to use this, or the info icon that was added in https://github.com/woocommerce/woocommerce/pull/37352? I think we should be consistent and use the info icon.

File: plugins/woocommerce/client/legacy/css/admin.scss (hunk 2 above)
Comment (mattsherman) [reply plugins/woocommerce/client/legacy/css/admin.scss:1069]: I turned off stylelint when I edited this file to avoid this massive change. Ideally we would do that linter fix in a separate PR (or commit at the very least), but what's done is done! 😄

Comment (mattsherman) [reply plugins/woocommerce/client/legacy/css/admin.scss:1069]: There is a lot of duplication between this style and the `.add-attributes-container` style added in #37352. Perhaps it would make sense to refactor a bit to avoid this duplication.

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js [hunk 3]
```diff
@@ -726,41 +739,51 @@ jQuery( function ( $ ) {
...
-					$( '#woocommerce-product-data' ).trigger(
-						'woocommerce_variations_loaded'
-					);
+						$( '#woocommerce-product-data' ).trigger(
+							'woocommerce_variations_loaded'
+						);
 
-					wc_meta_boxes_product_variations_ajax.unblock();
-				},
+						// wc_meta_boxes_product_variations_ajax.show_hide_variation_empty_state();
```
Comment (mattsherman) [inline plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js]: We should remove this commented out code!

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-variable-product.spec.js [hunk 4]
```diff
@@ -195,6 +198,9 @@ test.describe( 'Add New Variable Product Page', () => {
 		await page.click( 'a[href="#variable_product_options"]' );
 		await page.waitForLoadState( 'networkidle' );
 		await page.selectOption( '#field_to_edit', 'delete_all' );
+		await page.evaluate( () => {
+			window.confirm = () => true;
+		} );
```
Comment (mattsherman) [inline plugins/woocommerce/tests/e2e-pw/tests/merchant/create-variable-product.spec.js]: Why is this needed? We should comment this, since it isn't clear.

This is mocking the `confirm` browser dialog, right? Why is this needed?

Would it make more sense to do...

```
page.on('dialog', dialog => dialog.accept());
```

From: https://playwright.dev/docs/dialogs

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js [hunk 5]
```diff
@@ -726,41 +739,51 @@ jQuery( function ( $ ) {
 		 * @param {Int} per_page (default: 10)
 		 */
 		load_variations: function ( page, per_page ) {
-			page = page || 1;
-			per_page =
-				per_page ||
-				woocommerce_admin_meta_boxes_variations.variations_per_page;
+			return new Promise( ( resolve, reject ) => {
```
Comment (mattsherman) [inline plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js:745]: ❤️

**Review by mattsherman — DISMISSED** [review]
Love the use of Promise in the latest commits.

Tested, and it working well.

Left a few comments related to avoiding code duplication, and asking about the icon used.

Also, left a suggestion about Playwright and the `confirm` dialog.

This is really close -- great work so far!

File: plugins/woocommerce/assets/images/product_data/no-variation-background-image.svg
Comment (octaedro) [reply plugins/woocommerce/assets/images/product_data/no-variation-background-image.svg:1]: I think that they are referring to different things. The icon used in https://github.com/woocommerce/woocommerce/pull/37352 aims to show the user that there is information available in another place.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-variable-product.spec.js (hunk 4 above)
Comment (octaedro) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/create-variable-product.spec.js]: After rebasing my PR, there were introduced a few changes that broke that part and this is why I mocked the `confirm`. I realized that the rebase left a duplicated `on dialog` and removing one of them fixed the problem.

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js [hunk 6]
```diff
@@ -125,8 +125,9 @@ jQuery( function ( $ ) {
 		 * @param {Object} event
 		 * @param {Int} qty
 		 */
-		reload: function () {
-			wc_meta_boxes_product_variations_ajax.load_variations( 1 );
+		reload: async function () {
```
Comment (nathanss) [inline plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js]: Async/await is quite new. Do you know if we can use it here? I don't know whether this is transpiled or not

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js [hunk 7]
```diff
@@ -655,6 +656,18 @@ jQuery( function ( $ ) {
 				'a.do_variation_action',
 				this.do_variation_action
 			);
+
+			$( '.wc-metaboxes-wrapper' ).on(
```
Comment (nathanss) [inline plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js:663]: Shouldn't we attach to the parent? Do we still have a scenario where the contents of the variation screen can be replaced?

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js [hunk 8]
```diff
@@ -655,6 +656,18 @@ jQuery( function ( $ ) {
...
+				'click',
+				'button.generate_variations',
+				this.generate_variations
+			);
+
+			$( '.wc-metaboxes-wrapper' ).on(
```
Comment (nathanss) [inline plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js:669]: Same as above

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js [hunk 9]
```diff
@@ -1523,7 +1574,7 @@ jQuery( function ( $ ) {
 		/**
 		 * Paginav pagination selector
 		 */
-		page_selector: function () {
+		page_selector: async function () {
```
Comment (nathanss) [inline plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js]: Same question regarding async/await

**Review by nathanss — CHANGES_REQUESTED** [review]
image

Not touched by this PR, but: I find the terminology 'link' all variations confusing. The user just clicked a button called 'Generate variations', so I don't see why we're calling it 'link' here. It seems very technical. @jarekmorawski maybe we can improve this?

image
Is this the expected state after removing all the variations?

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js (hunk 6 above)
Comment (octaedro) [reply plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js]: The async/await was introduced in June 2017 (ECMAScript 2017 - ES8). I think that it makes sense to use `.then` instead (that was introduced a couple of years before). I did it in the commit [367d6f9](https://github.com/woocommerce/woocommerce/pull/37411/commits/367d6f99c15a8e47d5cf725612f6cc2ace63167d)

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js (hunk 7 above)
Comment (octaedro) [reply plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js:663]: I think that attaching the event to `.wc-metaboxes-wrapper` should be enough. That's the way the `Go` button has been attached as well. I couldn't find any edge case where we need something higher.

**Comment by octaedro** [comment]
Hey @nathanss,

image

> Is this the expected state after removing all the variations?

I couldn't reproduce that. Could you share the steps to see it?

**Review by nathanss — APPROVED** [review]

**Comment by jarekmorawski** [comment]
> The user just clicked a button called 'Generate variations', so I don't see why we're calling it 'link' here. It seems very technical.

I agree. Let's change it to `Do you want to generate all variations?`

> Is this the expected state after removing all the variations?

No. If you remove all variations, we should show the new empty state again (the one with the arrow).
